environments/python3/server.py

```diff
@@ -25,9 +25,11 @@ def loadv2():
     body = request.get_json()
     filepath = body['filepath']
     functionName = body['functionName']
-    # add filepath into syspath for module import
-    sys.path.append(filepath)
-    fn, path, desc = imp.find_module('user', [filepath])
+    # add module directory into syspath for module import
+    dirs = os.listdir(filepath)
+    for moduleDir in dirs:
```

Comment:
In previous fetcher `archive` function implementation, fetcher ignores the top level directory of deployment package and only archives the content of deployment package. When fetcher downloads & unarchives the package to `/userfunc` the directory structure would be like:

```
/userfunc/
├── user
│   ├── __init__.py
│   ├── build.sh
│   ├── requirements.txt
│   ├── user.py
│   └── <dependency files>
└── user.tmp
```

With this directory structure, python environment can simply append filepath to sys.path and load user function from it.

However, with the latest changes to `archive` function (archive top-level directory directly instead of its content), the directory structure becomes like:

```
/userfunc/
├── user
│   └── python-srcbuild-1507541251-qhnywl-n0herk-tj4tt8
│       ├── __init__.py
│       ├── build.sh
│       ├── requirements.txt
│       ├── user.py
│       └── <dependency files>
└── user.tmp
```

The directory name `python-srcbuild-1507541251-qhnywl-n0herk-tj4tt8` is generated randomly, python environment has no way to know the exactly directory name to add with. And since `python-srcbuild-1507541251-qhnywl-n0herk-tj4tt8` is the only directory under `/userfunc/user`, so here I let python env uses `os.listdir` to get directory name and add it to sys.path. If environment does not add directory to sys.path we may encounter the issue `ImportError: No module named 'user'`.

If we want to avoid this we can change the name `srcPkgFilename` in https://github.com/fission/fission/blob/master/buildermgr/common.go#L84
and the name `deployPkgFilename` in https://github.com/fission/fission/blob/master/builder/builder.go#L99 to fixed pattern instead of generating randomly.

Pros:
1. fetcher always knows the right path to load function since it is given by poolmgr.

Cons:
1. builder need to check the existence of directories before starting the build process, and remove them if necessary. Also that users may not be able to login env builder to check what happened in  directory after the build failure since the directory is removed. (maybe I worry too much, since such use case should never happen?)

What do you think about this? @soamvasani

Reply:
> fetcher ignores the top level directory of deployment package and only archives the content of deployment package.

Sorry, I don't understand. What is the difference between "contents of the package" and "top level directory of the package"?

Reply:
The goal with unique names is to (a) allow prefetching of functions (b) allow multiple functions to be loaded in the same environment. So any solution to the problems you describe should not assume that there's only one function being fetched.

Reply:
Ok, update from offline conversation -- the previous zip file design was better, and having a random name inside the zip file makes things complicated.